[Docs] Translate yolov5_tutorial.md (#138)
open-mmlab/mmyolo: docs/en/user_guides/yolov5_tutorial.md

# YOLOv5: From Start to Deployment

## Environment Setup

Note: Since this repository uses OpenMMLab 2.0, please create a new conda virtual environment to prevent conflicts with your existing repositories and projects of OpenMMLab 1.0.

```shell
conda create -n open-mmlab python=3.8 -y
conda activate open-mmlab
conda install pytorch torchvision -c pytorch
# conda install pytorch torchvision cpuonly -c pytorch
pip install -U openmim
mim install mmengine
mim install "mmcv>=2.0.0rc1,<2.1.0"
mim install "mmdet>=3.0.0rc0,<3.1.0"
# for albumentations
git clone https://github.com/open-mmlab/mmyolo.git
cd mmyolo
# Install albumentations
pip install -r requirements/albu.txt
# Install MMYOLO
mim install -v -e .
# "-v" means verbose, or more output
# "-e" means install the project in editable mode, so any local modifications made to the code will take effect, eliminating the need to reinstall.
```

For more detailed information about environment configuration, please refer to [get_started](../get_started.md).

## Dataset Preparation

In this tutorial, the training dataset for MMYOLO is less than 40MB and is selected from the balloon dataset.

```shell

> python tools/misc/download_dataset.py  --dataset-name balloon --save-dir data --unzip
> python tools/dataset_converters/balloon2coco.py

Reviewer: ```suggestion
python tools/misc/download_dataset.py  --dataset-name balloon \
                                                                  --save-dir data \
                                                                  --unzip
python tools/dataset_converters/balloon2coco.py
```

```

> After executing the above command, the balloon dataset will be downloaded in the `data` folder with the converted format we need. The `train.json` and `val.json` are the annotation files; both are in the coco format.

Reviewer: ```suggestion
After executing the above command, the balloon dataset will be downloaded in the `data` folder with the converted format we need. The `train.json` and `val.json` are the annotation files, both are in the coco format.
```

> # YOLOv5: From Start to Deploy

Reviewer 2: From getting started to deployment with YOLOv5

## Environment Setup

Note: Since this repository uses OpenMMLab 2.0, please create a new conda virtual environment to prevent conflicts with your existing repositories and projects of OpenMMLab 1.0.

```shell
conda create -n open-mmlab python=3.8 -y
conda activate open-mmlab
conda install pytorch torchvision -c pytorch
# conda install pytorch torchvision cpuonly -c pytorch
pip install -U openmim
mim install mmengine
mim install "mmcv>=2.0.0rc1,<2.1.0"
mim install "mmdet>=3.0.0rc0,<3.1.0"
# for albumentations
git clone https://github.com/open-mmlab/mmyolo.git
cd mmyolo
# Install albumentations
pip install -r requirements/albu.txt
# Install MMYOLO
mim install -v -e .
# "-v" means verbose, or more output
# "-e" means install the project in editable mode, so any local modifications made to the code will take effect, eliminating the need to reinstall.
```

For more detailed information about environment configuration, please refer to [get_started](../get_started.md).

## Dataset Preparation

> In this tutorial, the training dataset for MMYOLO is less than 40MB and is selected from the balloon dataset.

Reviewer 3: In this tutorial, we provide the ballon dataset, which is less than 40MB, as the training dataset for MMYOLO.

```shell

> python tools/misc/download_dataset.py  --dataset-name balloon \

Reviewer 2: python tools/misc/download_dataset.py  --dataset-name balloon --save-dir data --unzip

                                                                  --save-dir data \
                                                                  --unzip
python tools/dataset_converters/balloon2coco.py
```

> After executing the above command, the balloon dataset will be downloaded in the `data` folder with the converted format we need. The `train.json` and `val.json` are the annotation files, both are in the coco format.

Reviewer 3: The `train.json` and `val.json` are the annotation files in the coco format.

```python
_base_ = './yolov5_s-v61_syncbn_fast_8xb16-300e_coco.py'

data_root = 'data/balloon/'

train_batch_size_per_gpu = 4
train_num_workers = 2

metainfo = {
    'CLASSES': ('balloon', ),
    'PALETTE': [
        (220, 20, 60),
    ]
}

train_dataloader = dict(
    batch_size=train_batch_size_per_gpu,
    num_workers=train_num_workers,
    dataset=dict(
        data_root=data_root,
        metainfo=metainfo,
        data_prefix=dict(img='train/'),
        ann_file='train.json'))

val_dataloader = dict(
    dataset=dict(
        data_root=data_root,
        metainfo=metainfo,
        data_prefix=dict(img='val/'),
        ann_file='val.json'))

test_dataloader = val_dataloader

val_evaluator = dict(ann_file=data_root + 'val.json')

test_evaluator = val_evaluator

model = dict(bbox_head=dict(head_module=dict(num_classes=1)))

default_hooks = dict(logger=dict(interval=1))
```

The above configuration is inherited from `./yolov5_s-v61_syncbn_fast_8xb16-300e_coco.py`, and `data_root`, `metainfo`, `train_dataloader`, `val_dataloader`, `num_classes` and other configurations are updated according to the balloon data we are using.

> We set the `interval` of the logger to 1 because each iteration of the `interval` will output a loss-related log, and the balloon dataset we use is relatively tiny. We will not see the loss-related output if the `interval` is too large.

Reviewer 3: ```suggestion
The reason why we set the `interval` of the logger to 1 is that the balloon data set we choose is relatively small, and if the `interval` is too large, we will not see the output of the loss-related log. Therefore, by setting the `interval` of the logger to 1 will ensure that each interval iteration will output a loss-related log.
```
You can decide by your own.

## Training

```shell
python tools/train.py configs/yolov5/yolov5_s-v61_syncbn_fast_1xb4-300e_balloon.py
```

After executing the above training command, the `work_dirs/yolov5_s-v61_syncbn_fast_1xb4-300e_balloon` folder will be automatically generated. Both the weight and the training configuration files will be saved in this folder.

<div align=center>
<img src="https://cdn.vansin.top/img/20220913213846.png" alt="image"/>
</div>

### Resume training after interruptions

If training stops midway, add `--resume` at the end of the training command, and the program will automatically load the latest weight file from `work_dirs` to resume training.

```shell
python tools/train.py configs/yolov5/yolov5_s-v61_syncbn_fast_1xb4-300e_balloon.py --resume
```

### Fine-tune with loaded pretrained weights

> Our testing shows that with the YOLOv5-s pretrained model to train and validate, the coco/bbox_mAP can increase by more than 30 percent on the balloon dataset compared with not using the pre-training model solution.

Reviewer 3: NOTICE: It is highly recommended that finetuning from large datasets, such as COCO, can significantly boost the performance of overall network. 
In this example, compared with training from scratch, finetuning the pretrained model outperforms with a significant margin. (Over 30+ mAP boost than training from scratch).

```

Note: Ideally, the `strict_load` initialization parameter of `EMAHook` should be set to `False` during the fine-tuning phase, which makes the command `custom_hooks.0.strict_load=False`. However, because MMEngine v0.1.0 is the initial development version, there will be problems with this setting right now. Therefore, for the time being, you can only use the command `custom_hooks=None` to turn off `custom_hooks` to load the pre-training weights correctly. This issue is expected to be fixed in the next release.

3. Freeze backbone to train

Freeze the four backbone stages by setting `model.backbone.frozen_stages=4` in the config file or from the command line.

```shell
# Set model.backbone.frozen_stages=4 from the command line
cd mmyolo
python tools/train.py configs/yolov5/yolov5_s-v61_syncbn_fast_1xb4-300e_balloon.py \
                      --cfg-options load_from='yolov5_s-v61_syncbn_fast_8xb16-300e_coco_20220918_084700-86e02187.pth' model.backbone.frozen_stages=4 custom_hooks=None
```

### Visualization

For `visualization` of `default_hooks` in `configs/yolov5/yolov5_s-v61_syncbn_fast_1xb4-300e_balloon.py`, we set `draw` to `True` and `interval` to `2`.

> ```shell

Reviewer: ```suggestion
```python
```

default_hooks = dict(
     logger=dict(interval=1),
     visualization=dict(draw=True, interval=2),
)
```

> Then re-run the following training command. During the validation, each `interval` image will save a puzzle of the annotation and prediction results to `work_dirs/yolov5_s-v61_syncbn_fast_1xb4-300e_balloon/{timestamp}/vis_data/vis_image` folder.

Reviewer 3: Re-run

```

Re-run the training command, a Tensorboard folder will be created in `work_dirs/yolov5_s-v61_syncbn_fast_1xb4-300e_balloon/{timestamp}/vis_data`, You can get data visualization results such as loss, learning rate, and coco/bbox_mAP in the web link prompted on the command line with the following command:

```shell
tensorboard --logdir=work_dirs/yolov5_s-v61_syncbn_fast_1xb4-300e_balloon
```

## Model Testing

If you set `custom_hooks=None` during training, you still need to set `custom_hooks=None` during model testing as well.

```shell
python tools/test.py configs/yolov5/yolov5_s-v61_syncbn_fast_1xb4-300e_balloon.py \
                      work_dirs/yolov5_s-v61_syncbn_fast_1xb4-300e_balloon/epoch_300.pth \
                      --show-dir show_results --cfg-options custom_hooks=None
```

If you don't set `custom_hooks=None` during training, the test command should be as follows:

```shell
python tools/test.py configs/yolov5/yolov5_s-v61_syncbn_fast_1xb4-300e_balloon.py \
                      work_dirs/yolov5_s-v61_syncbn_fast_1xb4-300e_balloon/epoch_300.pth \
                      --show-dir show_results
```

> Run the above command, the inference result picture will be automatically saved to the `work_dirs/yolov5_s-v61_syncbn_fast_1xb4-300e_balloon/{timestamp}/show_results` folder. The following is one of the result pictures. Left picture is the actual annotation, and the right picture is the model inference result.

Reviewer 3: The left and the right
